Rhmap 16997 template apps refactor

mikenairn opened the pull request:
* Various updates, comments inline

mikenairn committed:
Add clone options to checkoutGitRepo (noTags, shallow)

mikenairn committed:
Use shallow clone on branchRepo task

mikenairn committed:
Remove sshagent and git config setup from branchRepo and tagRepo helpers (Slows things down when running multiple commands in parallel)

mikenairn committed:
Add type to component configs

mikenairn committed:
Add getTemplateAppComponentConfigs and writeTemplateAppConfigsGruntTask helpers

mikenairn committed:
Add component filter helpers

mikenairn committed:
Add Error handling in ghBranchProtectionApiRequest Helper

mikenairn committed:
Update getComponentConfigs helper to optionally try master git ref if checkout fails.

mikenairn committed:
Update branchRepo with option to force the use of any existing branch

mikenairn committed:
Extract baseBranch from template app config object.

mikenairn commented on src/org/feedhenry/GitHubUtils.groovy:
```diff
@@ -11,30 +11,33 @@ def ghApiRequest(String endpoint, String httpMethod = 'GET', String requestBody
     }
-    return branchStatus
+    return "${branchName} on ${ghRepo} in ${ghOrg} : ${branchStatus}"
 }
```
Prevent this method "ghBranchProtectionApiRequest" throwing an exception if the http request returns an unexpected status.

mikenairn commented on src/org/feedhenry/Utils.groovy:
```diff
@@ -49,3 +49,23 @@ static def getDate() {
+@NonCPS
+def filterPlatformComponents(components) {
+    components.findAll({ k, v -> v?.type == "platform" })
+}
```
Added these here for convenience, they were starting to get duplicated in a few places. Could probably be refactored further to provide a more generic filter helper.

mikenairn commented on vars/branchRepo.groovy:
```diff
@@ -1,42 +1,39 @@
         repoUrl = gitRepoUrl
         branch = gitRef
         targetDir = checkoutDir
+        shallow = true
```
Shallow clone to speed things up a bit.

mikenairn commented on vars/branchRepo.groovy:
```diff
@@ -1,42 +1,39 @@
 #!/usr/bin/groovy
 
-def call(String gitRepoUrl, String gitRef, String checkoutDir, String branchName, Closure body = {}) {
+def call(String gitRepoUrl, String gitRef, String checkoutDir, String branchName, boolean useExistingBranch = false, Closure body = {}) {
```
New optional argument, if true will use any exiting branch on the repo if it already exists, false it does the usual checks.

mikenairn commented on vars/checkoutGitRepo.groovy:
```diff
@@ -7,12 +7,15 @@ def call(body) {
     body()
 
+    def noTags = config.noTags  ?: false
+    def shallow = config.shallow  ?: false
```
Optionally pass these in to the checkout command.

mikenairn commented on vars/getComponentConfigs.groovy:
```diff
@@ -1,14 +1,27 @@
 import groovy.json.JsonSlurperClassic
 
-def call(componentConfigs, configGitRepo, configGitRef) {
+def call(componentConfigs, configGitRepo, configGitRef, tryMasterOnFail = false) {
```
New optional argument, if true will use the master branch to retrieve the component configs if the configGitRef passed in does not exist.

mikenairn commented on vars/getTemplateAppComponentConfigs.groovy:
```diff
@@ -0,0 +1,60 @@
+            }
+        }
+    }
+}
```
Used to clone the template apps repo (https://github.com/feedhenry/fh-template-apps) and return the template app configuration in the same format as getComponentConfigs.

```
def templateAppConfigs = [:]
getTemplateAppComponentConfigs(templateAppConfigs, git@github.com:feedhenry/fh-template-apps.git, 'master')
```

grdryn replied:
This method looks fine. Where possible though, I've got a preference for returning the output of functions, rather than passing in something as global state that's passed to the function as a param to be mutated (the `templateAppConfigs` map in your example).

Would this not be cleaner?

```groovy
def templateAppConfigs = getTemplateAppComponentConfigs('git@github.com:feedhenry/fh-template-apps.git', 'master')
```

Do we need to allow for the possibility that `templateAppConfigs` isn't empty going in to the function? We don't seem to read it from within the function, so if there's already values in it, we could concat the two maps after maybe?

mikenairn replied:
This is currently like this as that is how 'getComponentConfigs' was done initially. The usage of that has changed somewhat though and could probably be changed to just return a value now. Will give it a go as it would be nicer.

mikenairn commented on vars/tagRepo.groovy:
```diff
@@ -7,35 +7,30 @@ def call(String gitRepoUrl, String checkoutGitBranch, String checkoutDir, String
+            sh "git push origin ${tagName}"
         }
     }
 }
```
Remove sshagent and git config setup. This can be really slow if you are calling this method in a parallel step. Configure the node first and call this:

```
node('nodejs') {
        step([$class: 'WsCleanup'])
        ansiColor('xterm') {
            sshagent(['jenkinsgithub']) {
                sh "git config --global user.name \"Jenkins\""
                sh "git config --global user.email \"jenkins@wendy.feedhenry.net\""
                tagRepo(gitUrl, gitBranch, 'blah', gitTag, gitTag2)
        }
}
```

mikenairn commented on vars/writeTemplateAppsConfigGruntTask.groovy:
```diff
@@ -0,0 +1,59 @@
+};
+'''
+    }
+}
```
For use with getTemplateAppComponentConfigs.groovy, writes a temporary grunt task to generate the config file.

```
writeTemplateAppsConfigGruntTask()
sh('npm install --ignore-scripts')
sh('npm install grunt-cli -g')
sh("grunt configs --file-name 'TEMPLATE_APP_COMPONENTS.json'")
```

Will probably be moved into fh-template-apps the near future.

grdryn commented on src/org/feedhenry/GitHubUtils.groovy:
```diff
@@ -11,30 +11,33 @@ def ghApiRequest(String endpoint, String httpMethod = 'GET', String requestBody
             customHeaders: customHeaders,
-            url: url
+            url: url,
+            validResponseCodes: '100:399,404'
```
Is 404 always going to be considered a valid response for anyone using this `ghApiRequest` method?

mikenairn replied:
Probably not, i had meant to have this validResponsCodes string be passed in optionally like customHeaders

grdryn commented on vars/getTemplateAppComponentConfigs.groovy:
```diff
@@ -0,0 +1,60 @@
+            }
+        }
+
+        //ToDo Check this into template apps and remove this
```
Is this TODO expected to be done in this PR? If not, it might be good to create a JIRA issue in the FH project to get it done, and link that here in the comment. What do you think?

grdryn commented on vars/writeTemplateAppsConfigGruntTask.groovy:
```diff
@@ -0,0 +1,59 @@
+      if (_.has(obj, \'githubUrl\')) {
+        var templateAppConfig = {};
+        var parsedUrl = url.parse(obj.githubUrl, true, true);
+        var tarry = parsedUrl.path.replace(/.git$/, \'\').replace(/^\\//, \'\').split(\'/\');
```
What does tarry mean or stand for? :)

mikenairn replied:
template array, used to always name byte arrays barry, just a stupid habit.

grdryn commented on vars/writeTemplateAppsConfigGruntTask.groovy:
```diff
@@ -0,0 +1,59 @@
+    var templateAppConfigs = [];
+
+    this.filesSrc.forEach(function (filepath) {
+      updateTemplateAppConfig(filepath, templateAppConfigs);
```
Similar to the groovy example in `getTemplateAppComponentConfigs.groovy`, I think it could be better if we didn't pass in `templateAppConfigs` here to be modified within the function. Here's what I think you could have instead (after modifying the `updateTemplateAppConfig` function to just return the result rather than pushing it to `templateAppConfigs`):

```js
const templateAppConfigs = this.filesSrc.map(updateTemplateAppConfig);
```

What do you think?

Disclaimer: I haven't done any javascript in a while, but I think this should work

mikenairn replied:
This is based on what is already being done in template apps grunt tasks so will probably just leave it this way for now.

grdryn reviewed (commented):
@mikenairn this generally looks good. I've got some questions and comments inline.

grdryn commented:
I ran the job for this here: https://jenkins-wendy.ci.feedhenry.org/job/feedhenry/job/fh-pipeline-library/view/change-requests/job/PR-43/

I'm not sure why it didn't report on the status check.

Edit: it seems to be because the CI gh team hadn't been added to this repo. I've added it now and the check is being reported fine.

mikenairn committed:
Refactor config helpers (getComponentConfigs and getTemplateAppComponentConfigs) to return a components config object.
Move config file parsing logic into utils so it can be shared in all config helpers.

mikenairn committed:
Add optional argument for validResponseCodes in ghApiRequest.

mikenairn committed:
Update test Jenkinsfile:
* Update getComponentConfigs test stage
* Add getTemplateAppComponentConfigs test stage

grdryn reviewed (approved):
Looks good to me @mikenairn 👍
